metatensor-torch/src/tensor.cpp

```diff
@@ -15,7 +15,46 @@ bool custom_class_is(torch::IValue ivalue) {
     return ivalue.type().get() == expected_type;
 }
 
-TensorMapHolder::TensorMapHolder(metatensor::TensorMap tensor): tensor_(std::move(tensor)) {}
+TensorMapHolder::TensorMapHolder(metatensor::TensorMap tensor): tensor_(std::move(tensor)) {
+    const auto keys_count = this->tensor_.keys().count();
+    if (keys_count == 0) return;  // nothing to check
+
+    auto block_0 = this->tensor_.block_by_id(0);
+    const auto array_0 = block_0.mts_array();
+    const auto* ptr_0 = static_cast<metatensor::DataArrayBase*>(array_0.ptr);
+    const auto* torch_tensor_wrapper_0 = dynamic_cast<const TorchDataArray*>(ptr_0);
+    if (torch_tensor_wrapper_0 == nullptr) {
+        C10_THROW_ERROR(ValueError,
+            "this TensorBlock does not contain a C++ torch Tensor"
+        );
+    }
```

Comment:
I don't love that we need to access these internal details here. It might be worth to make this constructor private or even remove it entirely to prevent misusing it. I'll have a look at how much pain removing this would create.